fixes #297: fitting in batches without all labels

declanmillar opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Closes #297. Currently, VQC will throw an error if trained on batches of data where not all of the target labels that can be found in the full dataset are present. This is because VQC interprets the number of unique labels in the current batch as the number of classes. 

Currently, VQC is hard-coded to expect one-hot encoded labels. Therefore, I suggest a small change to determine the number of classes from the shape of the label array. I have also edited the docstring to add the label encoding requirements of the current VQC implementation.

### Details and comments

Related issues that I will raise separately:
- I may also add a warning when running with a warm start and fitting to data with a different number of one-hot labels, as I would expect this to invalidate any previous fit.
- While multi-label classification is not supported, the code will not raise an error when it is provided; it might be beneficial to raise an error in such cases.
- In future, it might be good for VQC to accept different label encodings, in line with what an sklearn user might expect.
- It may simplify the unit tests to add objects directly to the dtd data decorator, rather than using string tags, as discussed by @woodsp-ibm on #312.

woodsp-ibm commented:
Could we get a test case that would fail previously that will now work with this bug fix. Since its an aspect that CI did not catch it would be good to get a unit test in there to cover it.

Also don't forget the reno release note.

declanmillar commented:
>

declanmillar commented:
> Could we get a test case that would fail previously that will now work with this bug fix. Since its an aspect that CI did not catch it would be good to get a unit test in there to cover it.
> 
> Also don't forget the reno release note.

I've added a new unit test that would fail without this fix. It may be necessary to revise it if and when we implement a wider range of valid label encodings in the VQC interface. I've also added a reno note for the fix.

adekusar-drl reviewed (dismissed):
Looks good to me!
@woodsp-ibm Do you have open question?

dlasecki commented on test/algorithms/classifiers/test_vqc.py:
```diff
@@ -294,6 +294,61 @@ def test_warm_start(self, config):
+
+        if q_i == "statevector":
+            quantum_instance = self.sv_quantum_instance
+        else:
+            quantum_instance = self.qasm_quantum_instance
```
We might want to make this `if` explicit in case someone adds other backends to the test input in the future.

declanmillar replied:
Yep, seems sensible. Done.

dlasecki commented on qiskit_machine_learning/algorithms/classifiers/vqc.py:
```diff
@@ -155,12 +165,12 @@ def fit(self, X: np.ndarray, y: np.ndarray):  # pylint: disable=invalid-name
         Args:
             X: The input data.
-            y: The target values.
+            y: The target values. Required to be one-hot encoded.
```
Shall we add some input validation?

declanmillar replied:
I was thinking to open a new PR for input validation, but am happy to include it in this one!

woodsp-ibm replied:
Given this is a bug fix, and will be backported, we try to keep the code to what is needed to correct the critical bug for users in order to minimize code change and potentially destabilize code in the stable branch via other function such as new features. I think I might count validating the input - presumably checking all the labels as a new feature/improvement. If somehow you consider it vital its with this - otherwise I would consider it preferable on a separate PR that is not backported and would be part of next main release.

adekusar-drl replied:
I agree to introduce new input checks in a separate PR.

declanmillar replied:
Ok. I'll open a separate issue and PR for input validation.

woodsp-ibm commented on test/algorithms/classifiers/test_vqc.py:
```diff
@@ -294,6 +294,61 @@ def test_warm_start(self, config):
+        classifier.fit(features[5:10, :], target_onehot[5:10])
+
+        # Fit the VQC to the third third of the data with a warm start.
+        classifier.fit(features[10:, :], target_onehot[10:])
```
Is there anything we can assert here to test the outcome is as expected.? I imagine this test gets to the end and completes now rather than fails as it would have before the fix - so in that regard the outcome is implicitly tested. But it would be nice if something could be asserted expllicitly to better assess correctness.

declanmillar replied:
Yes, good point. I will add an assertion that the number of classes assumed by VQC is the same for each batch.

adekusar-drl commented:
@woodsp-ibm Is there anything left on your radar? If not, I'll approve and merge this PR.

woodsp-ibm reviewed (approved):
> Is there anything left on your radar? If not, I'll approve and merge this PR.
All looks good to me. May want to create issues for the other improvements/changes that are noted in the main comment so as we don't lose them.

adekusar-drl commented:
@declanmillar could you please post an issue(s) mentioned in this PR?
